core/authority-discovery/primitives/src/lib.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime Api to help discover authorities.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use client::decl_runtime_apis;
+use codec::Codec;
+use rstd::vec::Vec;
+
+decl_runtime_apis! {
+	/// The authority discovery api.
+	///
+	/// This api is used by the `core/authority-discovery` module to retrieve our
+	/// own authority identifier, to retrieve identifiers of the current authority
+	/// set, as well as sign and verify Kademlia Dht external address payloads
+	/// from and to other authorities.
+	///
+	/// # Note
+	///
+	/// The underlying cryptography key can change at any moment in time. This can
+	/// for example be problematic when a key change happens in between calling
+	/// `public_key` and `sign`, resulting in a signature that is not signed by
```

Comment:
This is not correct. If you use the same blockid for both calls, nothing will change.

Reply:
Good point. Also with having `sign` take an `authority_id` (see https://github.com/paritytech/substrate/pull/3385#discussion_r315278623) this race is removed.

Reply:
Hmm. Where is the authority_id stored between the calls?

Reply:
My process looks like this:

1. Retrieve authority id.

2. Prepare payload (possibly including the authority id) and serialize it.

3. Sign the payload with the private key corresponding to the authority id retrieved in *1*.

4. Put the *(key, value)* pair *(authority id, payload)* on the Dht.

> Hmm. Where is the authority_id stored between the calls?

It is not stored long term within the *authority-discovery* process.